### PR TITLE
On-each-release tap dance function

### PR DESCRIPTION
**Pull request by leep-frog**
## Description

The current tap dance logic only provides a hook for every key *tap*, but not for every key *release* in the dance.  This PR adds another hook function that is called for every key release in the tap dance.  It is only enabled when the `TAP_DANCE_ON_EACH_RELEASE` variable is defined in `config.h` (which will prevent any effects on current users).

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [X] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by leep-frog** [comment]
This CR is still ready for review.  What is the best way to get eyes on it?

File: quantum/process_keycode/process_tap_dance.h
Comment (drashna) [inline quantum/process_keycode/process_tap_dance.h:62]: This is ... messy, at best. 

If anything, I'd say that this should just have the on_each_release option, and just have the ifdef in the function (as it is in the c file).

Comment (leep-frog) [reply quantum/process_keycode/process_tap_dance.h]: Yea, I was thinking of the same approach initially, but I was trying to ensure that this feature didn't affect the memory usage for keyboards that didn't use that functionality, hence why I went with this implementation instead. 

But, as you mentioned, the implementation turned out less than ideal; in the newest commit, I removed the macro-conditionals such that the new function parameter is always present.  There's no need for the `TAP_DANCE_ON_EACH_RELEASE` anymore (since the function added in the `c` file will check for `NULL`).

Thanks for the initial review and comments, and let me know what you think of the changes.

File: quantum/process_keycode/process_tap_dance.c
Comment (leep-frog) [inline quantum/process_keycode/process_tap_dance.c:92]: This function checks if `on_each_release` is null btw (see [here](https://github.com/qmk/qmk_firmware/blob/fe98cc5dad538d137ec0484fbda282ef02c1156f/quantum/process_keycode/process_tap_dance.c#L76))

File: quantum/process_keycode/process_tap_dance.h
Comment (leep-frog) [inline quantum/process_keycode/process_tap_dance.h:44]: I added this at the end so existing user code that uses positional initialization doesn't break.  For example, the following will still compile even though only three of the four arguments are provided:

```
{ .fn = {user_fn_on_each_tap, user_fn_on_dance_finished, user_fn_on_dance_reset}, ... }
```

**Comment by leep-frog** [comment]
@drashna I just added tests, but the CI build is failing.  Seems like an issue with merging from the `main` branch into the `develop` branch though.  What's the best way to fix that?

**Comment by sigprof** [comment]
Apparently you merged something wrong into your branch.  Undo that merge by resetting to the last good commit:
```sh
git reset --hard a607f962086fed74b29e9cd3950fec8765a66b9c
```
Then try merging with the actual state of the `develop` branch:
```sh
git fetch upstream
git merge upstream/develop
```
The automatic merge would fail, because there were some other changes to `quantum/process_keycode/process_tap_dance.c` and `quantum/process_keycode/process_tap_dance.h`, so you would need to resolve the conflicts manually.  Finally, commit and push with `--force-with-lease`.

**Review by drashna — APPROVED** [review]

File: docs/feature_tap_dance.md
Comment (sigprof) [inline docs/feature_tap_dance.md:31]:
```suggestion
* `ACTION_TAP_DANCE_FN_ADVANCED_WITH_RELEASE(on_each_tap_fn, on_each_release_fn, on_dance_finished_fn, on_dance_reset_fn)`: This macro is identical to `ACTION_TAP_DANCE_FN_ADVANCED` with the addition of `on_each_release_fn` which is invoked every time the key for the tap dance is released. It is worth noting that `on_each_release_fn` will still be called even when the key is released after the dance finishes (e.g. if the key is released after being pressed and held for longer than the `TAPPING_TERM`).
```

**Review by sigprof — CHANGES_REQUESTED** [review]

**Comment by leep-frog** [comment]
@sigprof suggestion accepted.  Back to you

**Comment by leep-frog** [comment]
@sigprof @drashna can one of you merge this pull request?  I don't have write access and can't merge it.  Feel free to squash the commits as well when merging.

File: quantum/process_keycode/process_tap_dance.c
Comment (sigprof) [inline quantum/process_keycode/process_tap_dance.c:167]: I _suspect_ that the following change would be needed to fix the corner case when the `on_each_release` callback does `state->finished = true;` to finish the tap dance early (this kind of usage currently works for `on_each_tap` and is somewhat documented).
```suggestion
                    process_tap_dance_action_on_reset(action);
                    if (active_td == keycode) {
                        active_td = 0;
                    }
```
But I don't have time to test this just now (this would probably need another test to make sure that there are no superfluous `finished` calls after the tap dance timer expires).

And using `reset_tap_dance(state);` in `on_each_release()` would work properly anyway, because it resets `active_td` and clears the whole `state` (so the duplicate call of the `reset` callback won't happen); but it's not equivalent to `state->finished = true;` when used in the `on_each_tap` callback, so probably both cases should be supported.

Comment (leep-frog) [reply quantum/process_keycode/process_tap_dance.c:167]: I just added your change and another test function that verifies the expected behavior.  Let me know if there are any other cases you think should be added.

**Comment by leep-frog** [comment]
@sigprof I made changes for the comment you made earlier.  Mind re-reviewing and/or merging?

**Comment by leep-frog** [comment]
@sigprof @drashna Any chance either of you can merge this?

**Pull request merged**
